# Classify HFpEF

In [1]:
import pandas as pd
import numpy as np

from preproc import get_hfpef_200, convert2np

In [2]:
lbl_colname= ['Death', 'CV death', 'Major cardiac events']

## Clean Data

In [3]:
data_df = get_hfpef_200()

In [4]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 49 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Age                   200 non-null    int64  
 1   Sex                   200 non-null    int64  
 2   Cr                    200 non-null    float64
 3   GFR                   200 non-null    float64
 4   CKD stage             200 non-null    int64  
 5   smoke                 200 non-null    float64
 6   BMI                   200 non-null    float64
 7   BSA                   200 non-null    float64
 8   DM                    200 non-null    int64  
 9   Insulin               200 non-null    int64  
 10  AF                    200 non-null    int64  
 11  SBP                   200 non-null    int64  
 12  DBP                   200 non-null    int64  
 13  MAP                   200 non-null    float64
 14  PP                    200 non-null    int64  
 15  1/2SBP                2

In [5]:
data_df.describe()

,Age,Sex,Cr,GFR,CKD stage,smoke,BMI,BSA,DM,Insulin,...,E/A,Mitral E/e',TR Vmax,RWT,LV mass index,LVH,LAVI,LACI,LA diameter,LVEF
count,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,...,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000,200.000000
mean,72.045000,1.680000,1.711100,52.293250,2.810000,0.105000,26.738830,1.632980,0.560000,0.215000,...,0.741571,17.300650,1.890100,0.648305,115.640950,2.640000,31.987000,3.211636,3.670264,61.434500
std,11.404529,0.467647,1.534721,25.188572,1.043718,0.307323,6.636857,0.228277,0.497633,0.411853,...,0.606755,8.602117,1.443655,0.388423,52.778994,1.896943,26.285226,3.988946,1.446008,8.420635
min,31.000000,1.000000,0.550000,2.000000,1.000000,0.000000,15.400000,1.120000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,65.000000,1.000000,0.887500,34.750000,2.000000,0.000000,22.200000,1.485000,0.000000,0.000000,...,0.500000,11.600000,0.000000,0.400000,86.725000,1.000000,0.000000,0.000000,3.500000,56.000000
50%,74.000000,2.000000,1.140000,54.000000,3.000000,0.000000,26.300000,1.600000,1.000000,0.000000,...,0.700000,16.300000,2.460000,0.555000,113.150000,2.500000,32.600000,2.479869,4.000000,61.000000
75%,80.000000,2.000000,1.792500,70.250000,3.000000,0.000000,29.925000,1.800000,1.000000,0.000000,...,1.000000,22.100000,3.002500,0.865000,141.305000,4.000000,48.425000,5.111374,4.400000,66.000000
max,97.000000,2.000000,10.000000,108.000000,5.000000,1.000000,55.900000,2.300000,1.000000,1.000000,...,4.100000,42.200000,4.550000,2.300000,356.600000,7.000000,140.800000,23.665480,6.200000,82.900000


## Get Cleaned Data

In [6]:
data_df_nomiss = get_hfpef_200(nomiss=True)

In [8]:
data_df_nomiss.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 70 entries, 1 to 199
Data columns (total 49 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Age                   70 non-null     int64  
 1   Sex                   70 non-null     int64  
 2   Cr                    70 non-null     float64
 3   GFR                   70 non-null     float64
 4   CKD stage             70 non-null     int64  
 5   smoke                 70 non-null     float64
 6   BMI                   70 non-null     float64
 7   BSA                   70 non-null     float64
 8   DM                    70 non-null     int64  
 9   Insulin               70 non-null     int64  
 10  AF                    70 non-null     int64  
 11  SBP                   70 non-null     int64  
 12  DBP                   70 non-null     int64  
 13  MAP                   70 non-null     float64
 14  PP                    70 non-null     int64  
 15  1/2SBP                70

In [9]:
data_df_nomiss.describe()

,Age,Sex,Cr,GFR,CKD stage,smoke,BMI,BSA,DM,Insulin,...,E/A,Mitral E/e',TR Vmax,RWT,LV mass index,LVH,LAVI,LACI,LA diameter,LVEF
count,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,...,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000,70.000000
mean,72.014286,1.671429,1.704714,54.063429,2.728571,0.071429,26.173412,1.616043,0.628571,0.242857,...,0.977714,19.724143,1.961571,0.692429,121.555714,2.914286,40.508571,5.888278,4.000261,62.268571
std,11.643861,0.473085,1.557876,27.682150,1.141083,0.259399,6.245556,0.197336,0.486675,0.431906,...,0.482007,9.097155,1.387045,0.393100,46.882241,1.847386,18.354194,4.187923,0.672115,7.270798
min,31.000000,1.000000,0.640000,2.000000,1.000000,0.000000,15.400000,1.200000,0.000000,0.000000,...,0.400000,4.290000,0.000000,0.230000,53.800000,0.000000,12.500000,1.695652,2.100000,50.000000
25%,65.000000,1.000000,0.870000,32.000000,2.000000,0.000000,21.950000,1.455000,0.000000,0.000000,...,0.700000,13.050000,0.000000,0.420750,90.200000,1.000000,27.525000,3.142256,3.600000,57.475000
50%,74.500000,2.000000,1.120000,57.500000,3.000000,0.000000,25.450000,1.600000,1.000000,0.000000,...,0.900000,17.950000,2.440000,0.573500,113.450000,3.000000,36.900000,4.654273,4.000000,61.000000
75%,79.000000,2.000000,1.900000,76.500000,3.000000,0.000000,29.175000,1.775000,1.000000,0.000000,...,1.200000,23.500000,3.050000,0.900000,137.200000,4.000000,48.275000,6.919496,4.389250,67.000000
max,92.000000,2.000000,10.000000,108.000000,5.000000,1.000000,44.400000,2.000000,1.000000,1.000000,...,3.200000,42.200000,4.550000,2.300000,356.600000,7.000000,107.700000,23.665480,6.000000,78.200000


## Random Forest

In [10]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier

from sklearn.model_selection import cross_val_score, cross_val_predict, cross_validate 
from sklearn.metrics import classification_report

In [11]:
def hfpef_rf(X, y, feature_list):
    # Set the parameters by cross-validation
    tuned_parameters = [{'max_depth': [5,10,20],
                         'n_estimators': [100, 200, 400]}]

    rf = RandomForestClassifier(random_state=0)

    for i in range(y.shape[1]):
        print('==================================================')
        print('Prediction of ' + lbl_colname[i])

        y_true = y[:,i]

        clf = GridSearchCV(rf, tuned_parameters, scoring='f1_macro', cv=5)
        clf.fit(X, y_true)

        print("Best parameters set found on development set:")
        print()
        print(clf.best_params_)
        print()
        print("Grid scores on development set:")
        print()
        means = clf.cv_results_['mean_test_score']
        stds = clf.cv_results_['std_test_score']
        for mean, std, params in zip(means, stds, clf.cv_results_['params']):
            print("%0.3f (+/-%0.03f) for %r"
                  % (mean, std * 2, params))
        print()

        y_pred = cross_val_predict(clf.best_estimator_, X, y_true, cv=5)

        print(f"count {sum(y_pred)}/{len(y_pred)}")
        print("Detailed classification report:")
        print()
        print(classification_report(y_true, y_pred))
        print()

        for i,v in sorted(zip(feature_list, clf.best_estimator_.feature_importances_), key=lambda tup: tup[1], reverse=True)[:20]:
            print('Feature: %s, Score: %.5f' % (i,v))

### All Features

In [12]:
X, y, feature_list = convert2np(data_df, lbl_colname)

In [13]:
hfpef_rf(X, y, feature_list)

Prediction of Death
Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.458 (+/-0.006) for {'max_depth': 5, 'n_estimators': 100}
0.458 (+/-0.006) for {'max_depth': 5, 'n_estimators': 200}
0.458 (+/-0.006) for {'max_depth': 5, 'n_estimators': 400}
0.456 (+/-0.012) for {'max_depth': 10, 'n_estimators': 100}
0.458 (+/-0.006) for {'max_depth': 10, 'n_estimators': 200}
0.458 (+/-0.006) for {'max_depth': 10, 'n_estimators': 400}
0.456 (+/-0.012) for {'max_depth': 20, 'n_estimators': 100}
0.458 (+/-0.006) for {'max_depth': 20, 'n_estimators': 200}
0.458 (+/-0.006) for {'max_depth': 20, 'n_estimators': 400}

count 0/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.84      1.00      0.92       169
           1       0.00      0.00      0.00        31

    accuracy                           0.84       200
   macro avg       0.42      0.50      0.46       200
weight

C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.486 (+/-0.005) for {'max_depth': 5, 'n_estimators': 100}
0.486 (+/-0.005) for {'max_depth': 5, 'n_estimators': 200}
0.486 (+/-0.005) for {'max_depth': 5, 'n_estimators': 400}
0.486 (+/-0.005) for {'max_depth': 10, 'n_estimators': 100}
0.486 (+/-0.005) for {'max_depth': 10, 'n_estimators': 200}
0.486 (+/-0.005) for {'max_depth': 10, 'n_estimators': 400}
0.486 (+/-0.005) for {'max_depth': 20, 'n_estimators': 100}
0.486 (+/-0.005) for {'max_depth': 20, 'n_estimators': 200}
0.486 (+/-0.005) for {'max_depth': 20, 'n_estimators': 400}

count 0/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.94      1.00      0.97       189
           1       0.00      0.00      0.00        11

    accuracy                           0.94       200
   macro avg       0.47      0.50      0.49       200
weighted avg       0.89   

C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 200}

Grid scores on development set:

0.651 (+/-0.091) for {'max_depth': 5, 'n_estimators': 100}
0.701 (+/-0.068) for {'max_depth': 5, 'n_estimators': 200}
0.677 (+/-0.074) for {'max_depth': 5, 'n_estimators': 400}
0.646 (+/-0.072) for {'max_depth': 10, 'n_estimators': 100}
0.673 (+/-0.041) for {'max_depth': 10, 'n_estimators': 200}
0.677 (+/-0.047) for {'max_depth': 10, 'n_estimators': 400}
0.646 (+/-0.056) for {'max_depth': 20, 'n_estimators': 100}
0.683 (+/-0.037) for {'max_depth': 20, 'n_estimators': 200}
0.676 (+/-0.064) for {'max_depth': 20, 'n_estimators': 400}

count 108/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.72      0.67      0.69        99
           1       0.69      0.74      0.72       101

    accuracy                           0.70       200
   macro avg       0.71      0.70      0.70       200
weighted avg       0.71 

### Selected Features

In [14]:
X, y, feature_list = convert2np(data_df, lbl_colname, selected_feat=True)

In [15]:
hfpef_rf(X, y, feature_list)

Prediction of Death
Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.458 (+/-0.006) for {'max_depth': 5, 'n_estimators': 100}
0.458 (+/-0.006) for {'max_depth': 5, 'n_estimators': 200}
0.456 (+/-0.012) for {'max_depth': 5, 'n_estimators': 400}
0.455 (+/-0.018) for {'max_depth': 10, 'n_estimators': 100}
0.453 (+/-0.025) for {'max_depth': 10, 'n_estimators': 200}
0.453 (+/-0.025) for {'max_depth': 10, 'n_estimators': 400}
0.453 (+/-0.025) for {'max_depth': 20, 'n_estimators': 100}
0.453 (+/-0.025) for {'max_depth': 20, 'n_estimators': 200}
0.453 (+/-0.025) for {'max_depth': 20, 'n_estimators': 400}

count 0/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.84      1.00      0.92       169
           1       0.00      0.00      0.00        31

    accuracy                           0.84       200
   macro avg       0.42      0.50      0.46       200
weight

C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.486 (+/-0.005) for {'max_depth': 5, 'n_estimators': 100}
0.486 (+/-0.005) for {'max_depth': 5, 'n_estimators': 200}
0.486 (+/-0.005) for {'max_depth': 5, 'n_estimators': 400}
0.486 (+/-0.005) for {'max_depth': 10, 'n_estimators': 100}
0.486 (+/-0.005) for {'max_depth': 10, 'n_estimators': 200}
0.486 (+/-0.005) for {'max_depth': 10, 'n_estimators': 400}
0.486 (+/-0.005) for {'max_depth': 20, 'n_estimators': 100}
0.486 (+/-0.005) for {'max_depth': 20, 'n_estimators': 200}
0.486 (+/-0.005) for {'max_depth': 20, 'n_estimators': 400}

count 0/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.94      1.00      0.97       189
           1       0.00      0.00      0.00        11

    accuracy                           0.94       200
   macro avg       0.47      0.50      0.49       200
weighted avg       0.89   

C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'max_depth': 20, 'n_estimators': 100}

Grid scores on development set:

0.659 (+/-0.089) for {'max_depth': 5, 'n_estimators': 100}
0.669 (+/-0.081) for {'max_depth': 5, 'n_estimators': 200}
0.683 (+/-0.041) for {'max_depth': 5, 'n_estimators': 400}
0.673 (+/-0.075) for {'max_depth': 10, 'n_estimators': 100}
0.677 (+/-0.054) for {'max_depth': 10, 'n_estimators': 200}
0.666 (+/-0.085) for {'max_depth': 10, 'n_estimators': 400}
0.686 (+/-0.061) for {'max_depth': 20, 'n_estimators': 100}
0.666 (+/-0.066) for {'max_depth': 20, 'n_estimators': 200}
0.677 (+/-0.052) for {'max_depth': 20, 'n_estimators': 400}

count 107/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.70      0.66      0.68        99
           1       0.68      0.72      0.70       101

    accuracy                           0.69       200
   macro avg       0.69      0.69      0.69       200
weighted avg       0.69

### Cleaned Data

In [16]:
X, y, feature_list = convert2np(data_df_nomiss, lbl_colname, selected_feat=True)

In [17]:
hfpef_rf(X, y, feature_list)

Prediction of Death
Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.470 (+/-0.020) for {'max_depth': 5, 'n_estimators': 100}
0.470 (+/-0.020) for {'max_depth': 5, 'n_estimators': 200}
0.470 (+/-0.020) for {'max_depth': 5, 'n_estimators': 400}
0.470 (+/-0.020) for {'max_depth': 10, 'n_estimators': 100}
0.470 (+/-0.020) for {'max_depth': 10, 'n_estimators': 200}
0.470 (+/-0.020) for {'max_depth': 10, 'n_estimators': 400}
0.470 (+/-0.020) for {'max_depth': 20, 'n_estimators': 100}
0.470 (+/-0.020) for {'max_depth': 20, 'n_estimators': 200}
0.470 (+/-0.020) for {'max_depth': 20, 'n_estimators': 400}

count 0/70
Detailed classification report:

              precision    recall  f1-score   support

           0       0.89      1.00      0.94        62
           1       0.00      0.00      0.00         8

    accuracy                           0.89        70
   macro avg       0.44      0.50      0.47        70
weighte

C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\model_select

Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.793 (+/-0.508) for {'max_depth': 5, 'n_estimators': 100}
0.793 (+/-0.508) for {'max_depth': 5, 'n_estimators': 200}
0.793 (+/-0.508) for {'max_depth': 5, 'n_estimators': 400}
0.793 (+/-0.508) for {'max_depth': 10, 'n_estimators': 100}
0.793 (+/-0.508) for {'max_depth': 10, 'n_estimators': 200}
0.793 (+/-0.508) for {'max_depth': 10, 'n_estimators': 400}
0.793 (+/-0.508) for {'max_depth': 20, 'n_estimators': 100}
0.793 (+/-0.508) for {'max_depth': 20, 'n_estimators': 200}
0.793 (+/-0.508) for {'max_depth': 20, 'n_estimators': 400}



C:\Users\HP\miniconda3\lib\site-packages\sklearn\model_selection\_split.py:666: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.
  warnings.warn(("The least populated class in y has only %d"


count 0/70
Detailed classification report:

              precision    recall  f1-score   support

           0       0.97      1.00      0.99        68
           1       0.00      0.00      0.00         2

    accuracy                           0.97        70
   macro avg       0.49      0.50      0.49        70
weighted avg       0.94      0.97      0.96        70


Feature: LA diameter, Score: 0.17221
Feature: GFR, Score: 0.11888
Feature: Cr, Score: 0.10925
Feature: LAVI, Score: 0.09696
Feature: PP, Score: 0.09530
Feature: medial E', Score: 0.08193
Feature: Age, Score: 0.07653
Feature: CKD stage, Score: 0.06690
Feature: MAP, Score: 0.06206
Feature: NTProBNP, Score: 0.05547
Feature: medial a', Score: 0.05378
Feature: Sex, Score: 0.00691
Feature: AF, Score: 0.00381
Prediction of Major cardiac events


C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'max_depth': 5, 'n_estimators': 100}

Grid scores on development set:

0.552 (+/-0.207) for {'max_depth': 5, 'n_estimators': 100}
0.501 (+/-0.284) for {'max_depth': 5, 'n_estimators': 200}
0.476 (+/-0.240) for {'max_depth': 5, 'n_estimators': 400}
0.503 (+/-0.140) for {'max_depth': 10, 'n_estimators': 100}
0.513 (+/-0.137) for {'max_depth': 10, 'n_estimators': 200}
0.464 (+/-0.197) for {'max_depth': 10, 'n_estimators': 400}
0.503 (+/-0.140) for {'max_depth': 20, 'n_estimators': 100}
0.503 (+/-0.140) for {'max_depth': 20, 'n_estimators': 200}
0.464 (+/-0.197) for {'max_depth': 20, 'n_estimators': 400}

count 52/70
Detailed classification report:

              precision    recall  f1-score   support

           0       0.50      0.33      0.40        27
           1       0.65      0.79      0.72        43

    accuracy                           0.61        70
   macro avg       0.58      0.56      0.56        70
weighted avg       0.59   

##  Neural Networks

In [18]:
from sklearn.neural_network import MLPClassifier
from sklearn import preprocessing

In [19]:
def hfpef_ann(X, y, feature_list):
    # Set the parameters by cross-validation
    tuned_parameters = [{'hidden_layer_sizes': [(20,20,20), (100),(20,20)]}]

    ann = MLPClassifier(activation='tanh', max_iter=1000, random_state=0)

    for i in range(y.shape[1]):
        print('==================================================')
        print('Prediction of ' + lbl_colname[i])

        y_true = y[:,i]

        clf = GridSearchCV(ann, tuned_parameters, scoring='f1_macro', cv=5)
        clf.fit(X, y_true)

        print("Best parameters set found on development set:")
        print()
        print(clf.best_params_)
        print()
        print("Grid scores on development set:")
        print()
        means = clf.cv_results_['mean_test_score']
        stds = clf.cv_results_['std_test_score']
        for mean, std, params in zip(means, stds, clf.cv_results_['params']):
            print("%0.3f (+/-%0.03f) for %r"
                  % (mean, std * 2, params))
        print()

        y_pred = cross_val_predict(clf.best_estimator_, X, y_true, cv=5)

        print(f"count {sum(y_pred)}/{len(y_pred)}")
        print("Detailed classification report:")
        print()
        print(classification_report(y_true, y_pred))
        print()

### All Features

In [20]:
X, y, feature_list = convert2np(data_df, lbl_colname)

from sklearn.decomposition import PCA

X = PCA(n_components=0.95).fit_transform(X)
# scaler = preprocessing.MinMaxScaler().fit(X)
# X = scaler.transform(X)

In [21]:
X.shape

(200, 1)

In [22]:
hfpef_ann(X, y, feature_list)

Prediction of Death
Best parameters set found on development set:

{'hidden_layer_sizes': 100}

Grid scores on development set:

0.458 (+/-0.006) for {'hidden_layer_sizes': (20, 20, 20)}
0.505 (+/-0.193) for {'hidden_layer_sizes': 100}
0.458 (+/-0.006) for {'hidden_layer_sizes': (20, 20)}

count 44/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.85      0.79      0.82       169
           1       0.18      0.26      0.21        31

    accuracy                           0.70       200
   macro avg       0.52      0.52      0.52       200
weighted avg       0.75      0.70      0.72       200


Prediction of CV death
Best parameters set found on development set:

{'hidden_layer_sizes': 100}

Grid scores on development set:

0.486 (+/-0.005) for {'hidden_layer_sizes': (20, 20, 20)}
0.489 (+/-0.154) for {'hidden_layer_sizes': 100}
0.486 (+/-0.005) for {'hidden_layer_sizes': (20, 20)}

count 44/200
Detailed classification repo

==================================================
Prediction of Death
Best parameters set found on development set:

{'hidden_layer_sizes': (20, 20, 20)}

Grid scores on development set:

0.519 (+/-0.125) for {'hidden_layer_sizes': (20, 20, 20)}
0.473 (+/-0.071) for {'hidden_layer_sizes': 100}
0.458 (+/-0.006) for {'hidden_layer_sizes': (20, 20)}

count 11/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.85      0.95      0.90       169
           1       0.27      0.10      0.14        31

    accuracy                           0.82       200
   macro avg       0.56      0.52      0.52       200
weighted avg       0.76      0.82      0.78       200


==================================================
Prediction of CV death
Best parameters set found on development set:

{'hidden_layer_sizes': 100}

Grid scores on development set:

0.485 (+/-0.007) for {'hidden_layer_sizes': (20, 20, 20)}
0.486 (+/-0.005) for {'hidden_layer_sizes': 100}
0.486 (+/-0.005) for {'hidden_layer_sizes': (20, 20)}

count 0/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.94      1.00      0.97       189
           1       0.00      0.00      0.00        11

    accuracy                           0.94       200
   macro avg       0.47      0.50      0.49       200
weighted avg       0.89      0.94      0.92       200


==================================================
Prediction of Major cardiac events

{'hidden_layer_sizes': (20, 20, 20)}

Grid scores on development set:

0.685 (+/-0.132) for {'hidden_layer_sizes': (20, 20, 20)}
0.661 (+/-0.076) for {'hidden_layer_sizes': 100}
0.666 (+/-0.103) for {'hidden_layer_sizes': (20, 20)}

count 111/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.71      0.64      0.67        99
           1       0.68      0.74      0.71       101

    accuracy                           0.69       200
   macro avg       0.69      0.69      0.69       200
weighted avg       0.69      0.69      0.69       200

### Selected Features

In [23]:
X, y, feature_list = convert2np(data_df, lbl_colname, selected_feat=True)

X = PCA(n_components=0.95).fit_transform(X)

# scaler = preprocessing.MinMaxScaler().fit(X)
# X = scaler.transform(X)

In [24]:
hfpef_ann(X, y, feature_list)

Prediction of Death
Best parameters set found on development set:

{'hidden_layer_sizes': 100}

Grid scores on development set:

0.458 (+/-0.006) for {'hidden_layer_sizes': (20, 20, 20)}
0.505 (+/-0.193) for {'hidden_layer_sizes': 100}
0.458 (+/-0.006) for {'hidden_layer_sizes': (20, 20)}

count 44/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.85      0.79      0.82       169
           1       0.18      0.26      0.21        31

    accuracy                           0.70       200
   macro avg       0.52      0.52      0.52       200
weighted avg       0.75      0.70      0.72       200


Prediction of CV death
Best parameters set found on development set:

{'hidden_layer_sizes': (20, 20, 20)}

Grid scores on development set:

0.486 (+/-0.005) for {'hidden_layer_sizes': (20, 20, 20)}
0.485 (+/-0.145) for {'hidden_layer_sizes': 100}
0.486 (+/-0.005) for {'hidden_layer_sizes': (20, 20)}

count 0/200
Detailed classificat

C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'hidden_layer_sizes': (20, 20, 20)}

Grid scores on development set:

0.603 (+/-0.135) for {'hidden_layer_sizes': (20, 20, 20)}
0.603 (+/-0.135) for {'hidden_layer_sizes': 100}
0.603 (+/-0.135) for {'hidden_layer_sizes': (20, 20)}

count 44/200
Detailed classification report:

              precision    recall  f1-score   support

           0       0.58      0.92      0.71        99
           1       0.82      0.36      0.50       101

    accuracy                           0.64       200
   macro avg       0.70      0.64      0.61       200
weighted avg       0.70      0.64      0.60       200




### Cleaned Data

In [25]:
X, y, feature_list = convert2np(data_df_nomiss, lbl_colname, selected_feat=True)

# scaler = preprocessing.StandardScaler().fit(X)
# X = scaler.transform(X)

In [26]:
hfpef_ann(X, y, feature_list)

Prediction of Death
Best parameters set found on development set:

{'hidden_layer_sizes': (20, 20, 20)}

Grid scores on development set:

0.470 (+/-0.020) for {'hidden_layer_sizes': (20, 20, 20)}
0.470 (+/-0.020) for {'hidden_layer_sizes': 100}
0.470 (+/-0.020) for {'hidden_layer_sizes': (20, 20)}

count 0/70
Detailed classification report:

              precision    recall  f1-score   support

           0       0.89      1.00      0.94        62
           1       0.00      0.00      0.00         8

    accuracy                           0.89        70
   macro avg       0.44      0.50      0.47        70
weighted avg       0.78      0.89      0.83        70


Prediction of CV death


C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\model_select

Best parameters set found on development set:

{'hidden_layer_sizes': (20, 20, 20)}

Grid scores on development set:

0.793 (+/-0.508) for {'hidden_layer_sizes': (20, 20, 20)}
0.793 (+/-0.508) for {'hidden_layer_sizes': 100}
0.793 (+/-0.508) for {'hidden_layer_sizes': (20, 20)}



C:\Users\HP\miniconda3\lib\site-packages\sklearn\model_selection\_split.py:666: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.
  warnings.warn(("The least populated class in y has only %d"


count 0/70
Detailed classification report:

              precision    recall  f1-score   support

           0       0.97      1.00      0.99        68
           1       0.00      0.00      0.00         2

    accuracy                           0.97        70
   macro avg       0.49      0.50      0.49        70
weighted avg       0.94      0.97      0.96        70


Prediction of Major cardiac events


C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\miniconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Best parameters set found on development set:

{'hidden_layer_sizes': 100}

Grid scores on development set:

0.380 (+/-0.027) for {'hidden_layer_sizes': (20, 20, 20)}
0.509 (+/-0.153) for {'hidden_layer_sizes': 100}
0.380 (+/-0.027) for {'hidden_layer_sizes': (20, 20)}

count 55/70
Detailed classification report:

              precision    recall  f1-score   support

           0       0.47      0.26      0.33        27
           1       0.64      0.81      0.71        43

    accuracy                           0.60        70
   macro avg       0.55      0.54      0.52        70
weighted avg       0.57      0.60      0.57        70


